src/shogun/mathematics/linalg/LinalgBackendBase.h

```diff
@@ -76,59 +77,124 @@ class LinalgBackendBase
 	BACKEND_GENERIC_DOT(complex128_t);
 	#undef BACKEND_GENERIC_DOT
 
+	/** @see linalg::sum */
+	#define BACKEND_GENERIC_SUM(Type, Class) \
+	virtual Type sum(const Class<Type>&) const \
+	{  \
+		SG_SNOTIMPLEMENTED; \
+	}
+
+	BACKEND_GENERIC_SUM(bool, SGVector);
+	BACKEND_GENERIC_SUM(char, SGVector);
+	BACKEND_GENERIC_SUM(int8_t, SGVector);
+	BACKEND_GENERIC_SUM(uint8_t, SGVector);
+	BACKEND_GENERIC_SUM(int16_t, SGVector);
+	BACKEND_GENERIC_SUM(uint16_t, SGVector);
+	BACKEND_GENERIC_SUM(int32_t, SGVector);
+	BACKEND_GENERIC_SUM(uint32_t, SGVector);
+	BACKEND_GENERIC_SUM(int64_t, SGVector);
+	BACKEND_GENERIC_SUM(uint64_t, SGVector);
+	BACKEND_GENERIC_SUM(float32_t, SGVector);
+	BACKEND_GENERIC_SUM(float64_t, SGVector);
+	BACKEND_GENERIC_SUM(floatmax_t, SGVector);
+	BACKEND_GENERIC_SUM(complex128_t, SGVector);
+	BACKEND_GENERIC_SUM(bool, SGMatrix);
+	BACKEND_GENERIC_SUM(char, SGMatrix);
+	BACKEND_GENERIC_SUM(int8_t, SGMatrix);
+	BACKEND_GENERIC_SUM(uint8_t, SGMatrix);
+	BACKEND_GENERIC_SUM(int16_t, SGMatrix);
+	BACKEND_GENERIC_SUM(uint16_t, SGMatrix);
+	BACKEND_GENERIC_SUM(int32_t, SGMatrix);
+	BACKEND_GENERIC_SUM(uint32_t, SGMatrix);
+	BACKEND_GENERIC_SUM(int64_t, SGMatrix);
+	BACKEND_GENERIC_SUM(uint64_t, SGMatrix);
+	BACKEND_GENERIC_SUM(float32_t, SGMatrix);
+	BACKEND_GENERIC_SUM(float64_t, SGMatrix);
+	BACKEND_GENERIC_SUM(floatmax_t, SGMatrix);
+	BACKEND_GENERIC_SUM(complex128_t, SGMatrix);
+	#undef BACKEND_GENERIC_SUM
```

Comment:
Can you please use something clever here to avoid writing these ptypes all the time? For example, something like the following should work. You can then just use the macro `DEFINE_FOR_ALL_PTYPE` to define all your methods for all the ptypes. It's cleaner and shorter (although does the same job under the hood).
```c++
#include <iostream>
#include <memory>

using std::cout;
using std::endl;
using std::make_unique;

typedef float float32_t;
typedef double float64_t;

template <typename T>
struct GPUMemoryBase
{
};

template <typename T>
struct SGVector
{
};

template <typename T>
struct SGMatrix
{
};

#define SG_SNOTIMPLEMENTED cout << "Method not implemented" << endl;

#define DEFINE_FOR_ALL_PTYPE(METHODNAME, Container) \
	METHODNAME(int32_t, Container) \
	METHODNAME(float32_t, Container) \
	METHODNAME(float64_t, Container)
	
struct LinalgBackendBase
{
	#define BACKEND_GENERIC_TO_GPU(Type, Container) \
	virtual GPUMemoryBase<Type>* to_gpu(const Container<Type>&) const \
	{  \
		SG_SNOTIMPLEMENTED; \
		return nullptr; \
	}
	
	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_TO_GPU, SGVector)
	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_TO_GPU, SGMatrix)
};

int main() {
	auto backend = make_unique<LinalgBackendBase>();
	
	SGVector<int32_t> a;
	auto gpu_vec = backend->to_gpu(a);
	
	SGMatrix<float64_t> b;
	auto gpu_mat = backend->to_gpu(b);
	
	return 0;
}
```